File: DoublyLinkedList.java

public class DoublyLinkedList {

    // Node class
    static class Node {

        int data;
        Node previous;
        Node next;

        // Constructor
        Node(int data) {
            this.data = data;
            this.previous = null;
            this.next = null;
        }

    }

    // Head points to the first node
    private Node head = null;

    // Tail points to the last node
    private Node tail = null;

    // Insert a new node at the end
    public void insert(int data) {

        // Create a new node
        Node newNode = new Node(data);

        // If the list is empty
        if (head == null) {
            head = newNode;
            tail = newNode;
            return;
        }

        // Connect the new node with the last node
        tail.next = newNode;
        newNode.previous = tail;

        // Update the tail
        tail = newNode;
    }

    // Display from head to tail
    public void displayForward() {

        if (head == null) {
            System.out.println("Doubly Linked List is empty.");
            return;
        }

        Node current = head;

        System.out.print("Forward: ");

        while (current != null) {

            System.out.print(current.data);

            if (current.next != null) {
                System.out.print(" <-> ");
            }

            current = current.next;
        }

        System.out.println();
    }

    // Display from tail to head
    public void displayBackward() {

        if (tail == null) {
            System.out.println("Doubly Linked List is empty.");
            return;
        }

        Node current = tail;

        System.out.print("Backward: ");

        while (current != null) {

            System.out.print(current.data);

            if (current.previous != null) {
                System.out.print(" <-> ");
            }

            current = current.previous;
        }

        System.out.println();
    }

    public static void main(String[] args) {

        // Create Doubly Linked List object
        DoublyLinkedList list = new DoublyLinkedList();

        // Insert elements
        list.insert(10);
        list.insert(20);
        list.insert(30);
        list.insert(40);

        // Display list
        list.displayForward();

        list.displayBackward();

    }

}

File: DoublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>

// Node Structure
struct Node {
    int data;
    struct Node *prev;
    struct Node *next;
};

struct Node *head = NULL;

// Insert at Beginning
void insertAtBeginning(int value) {

    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));

    newNode->data = value;
    newNode->prev = NULL;
    newNode->next = head;

    if(head != NULL) {
        head->prev = newNode;
    }

    head = newNode;

    printf("%d inserted at the beginning.\n", value);
}

// Insert at End
void insertAtEnd(int value) {

    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));

    newNode->data = value;
    newNode->next = NULL;

    if(head == NULL) {
        newNode->prev = NULL;
        head = newNode;

        printf("%d inserted as the first node.\n", value);
        return;
    }

    struct Node *temp = head;

    while(temp->next != NULL) {
        temp = temp->next;
    }

    temp->next = newNode;
    newNode->prev = temp;

    printf("%d inserted at the end.\n", value);
}

// Delete by Value
void deleteNode(int value) {

    if(head == NULL) {
        printf("Linked List is Empty.\n");
        return;
    }

    struct Node *temp = head;

    while(temp != NULL && temp->data != value) {
        temp = temp->next;
    }

    if(temp == NULL) {
        printf("%d not found.\n", value);
        return;
    }

    if(temp == head) {
        head = temp->next;

        if(head != NULL) {
            head->prev = NULL;
        }
    }
    else {

        temp->prev->next = temp->next;

        if(temp->next != NULL) {
            temp->next->prev = temp->prev;
        }
    }

    free(temp);

    printf("%d deleted successfully.\n", value);
}

// Forward Traversal
void displayForward() {

    if(head == NULL) {
        printf("Linked List is Empty.\n");
        return;
    }

    struct Node *temp = head;

    printf("Forward Traversal:\n");

    while(temp != NULL) {
        printf("%d <-> ", temp->data);
        temp = temp->next;
    }

    printf("NULL\n");
}

// Backward Traversal
void displayBackward() {

    if(head == NULL) {
        printf("Linked List is Empty.\n");
        return;
    }

    struct Node *temp = head;

    while(temp->next != NULL) {
        temp = temp->next;
    }

    printf("Backward Traversal:\n");

    while(temp != NULL) {
        printf("%d <-> ", temp->data);
        temp = temp->prev;
    }

    printf("NULL\n");
}

int main() {

    int choice, value;

    do {

        printf("\n====== DOUBLY LINKED LIST MENU ======\n");
        printf("1. Insert at Beginning\n");
        printf("2. Insert at End\n");
        printf("3. Delete\n");
        printf("4. Display Forward\n");
        printf("5. Display Backward\n");
        printf("6. Exit\n");

        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch(choice) {

            case 1:
                printf("Enter value: ");
                scanf("%d", &value);
                insertAtBeginning(value);
                break;

            case 2:
                printf("Enter value: ");
                scanf("%d", &value);
                insertAtEnd(value);
                break;

            case 3:
                printf("Enter value to delete: ");
                scanf("%d", &value);
                deleteNode(value);
                break;

            case 4:
                displayForward();
                break;

            case 5:
                displayBackward();
                break;

            case 6:
                printf("Program Ended.\n");
                break;

            default:
                printf("Invalid Choice!\n");
        }

    } while(choice != 6);

    return 0;
}
